### PR TITLE
Rewrite message hooks tests part2

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
This picks up where we left off on expunging mock testing, this time for (the chat portion) of `MessageWillBePosted`. In addition to bumping the server dependencies, introduce a new `runPermutations` generic that takes a `struct` of booleans and simplifies testing all the combinations. (Thanks @JulienTant for pushing me towards generics to make this cleaner!)

#### Ticket Link
Relates-to: https://mattermost.atlassian.net/browse/MM-55512

### CODE REVIEW

File: server/main_test.go [hunk 1]
```diff
@@ -52,19 +50,6 @@ func (mt *mainT) FailNow() {
 	os.Exit(1)
 }
 
-// setupServerPath defines MM_SERVER_PATH correctly for server initialization. One day this won't
```
Comment (lieut-data) [inline server/main_test.go:55]: No longer needed, now that `MM_SERVER_PATH` is no more 🎉.

File: server/message_hooks_test.go [hunk 2]
```diff
@@ -822,7 +821,299 @@ func TestReactionHasBeenRemoved(t *testing.T) {
 	})
 }
 
+func TestMessageHasBeenPosted(t *testing.T) {
```
Comment (lieut-data) [inline server/message_hooks_test.go:824]: This is the bulk of this change -- most other diffs are just files being split apart for clarity.

File: server/message_hooks_test.go [hunk 3]
```diff
@@ -822,7 +821,299 @@ func TestReactionHasBeenRemoved(t *testing.T) {
...
+	})
+
+	type parameters struct {
+		SenderConnected    bool
+		SelectiveSync      bool
+		SyncDirectMessages bool
+		SyncGroupMessages  bool
+		SyncLinkedChannels bool
+	}
+	runPermutations(t, parameters{}, func(t *testing.T, params parameters) {
```
Comment (lieut-data) [inline server/message_hooks_test.go:901]: Note the permutation setup.

File: server/message_hooks_test.go [hunk 4]
```diff
@@ -822,7 +821,299 @@ func TestReactionHasBeenRemoved(t *testing.T) {
...
+			user1 := th.SetupUser(t, team)
+
+			client1 := th.SetupClient(t, user1.Id)
+
+			channel, _, err := client1.CreateDirectChannel(context.TODO(), user1.Id, user1.Id)
+			require.NoError(t, err)
+
+			if params.SenderConnected {
+				th.ConnectUser(t, user1.Id)
+			}
```
Comment (lieut-data) [inline server/message_hooks_test.go:923]: And when testing, I can just use the permutation parameters to properly setup the test.

File: server/message_hooks_test.go [hunk 5]
```diff
@@ -822,7 +821,299 @@ func TestReactionHasBeenRemoved(t *testing.T) {
...
+			channel, _, err := client1.CreateGroupChannel(context.TODO(), []string{user1.Id, user2.Id, user3.Id})
+			require.NoError(t, err)
+
+			if params.SenderConnected {
+				th.ConnectUser(t, user1.Id)
+			}
+
+			// Unconditionally link user2 with their Teams account. This is actually
+			// a current shortcoming with GMs in that we can't even start a conversation
+			// until we know the mapping for all users involved.
```
[the remark this answers is not in the document]
Comment (lieut-data) [reply server/message_hooks_test.go:1079]: This particular situation is with something like `user1` connected, `user2` never connected, and `user3` remote -- if I start a conversation with that group, messages were just getting dropped until we knew something about `user2`.

**Review by sbishel — APPROVED** [review]
Wow. Nice.

**Comment by sbishel** [comment]
I think your conflict is just a variable name change I introduced. So, I think you can just remove the conflict code.
https://github.com/mattermost/mattermost-plugin-msteams/pull/652/files#diff-d581d8e52a46c862bbe83a49a791197b964a9df2c95fc94185925c6e44323116L535

**Pull request merged**